### PR TITLE
Demote stale synced experiments if files not on S3

### PR DESCRIPTION
**Pull request by milesAraya**
#### Summary
- A published experiment can be marked `local_sync_status = synced` yet be undisplayable because its output files were later removed from S3. Previously the public reproduce endpoint just returned a data error and left the row stuck at `synced` forever, so the background sync job never re-checked it.
- On a `synced` row that fails display validation, `public_reproduce_experiment` now confirms against S3 and, **only when the data is confirmed absent**, demotes the row to `error` so the background sync job picks it up on its next pass.
- A transient S3 check failure (network/throttle/creds) no longer demotes a healthy row — it is left as `synced` and retried later.

#### Design Decisions
- **Tri-state S3 check instead of a boolean.** `_validate_experiment_exists_in_s3` now returns `True` (confirmed present) / `False` (confirmed absent, `KeyCount == 0`) / `None` (couldn't check — exception). The demotion path and the publish path need opposite fail-safe directions: demotion must **not** fire on uncertainty (don't corrupt a healthy row), while publish must **block** on uncertainty (don't publish data we can't confirm). A single boolean cannot serve both, which is why the exception case previously caused false demotions.
- **Demotion guarded by `if exists is False`**, so only a confirmed-absent object triggers the status change. The publish caller keeps `if not exists:` — `not None` and `not False` are both truthy, so it still blocks publish on absent **and** unknown (behavior unchanged there).
- **Idempotent UPDATE.** The demotion `UPDATE` is scoped with `WHERE publish_status = on AND local_sync_status = synced`, so concurrent requests converge and a second demotion is a no-op.
- **Rejected:** re-downloading from S3 on this path (the row is already `synced`, so `needs_sync` is false by design — the real problem is upstream data loss, not a local miss), and changing the publish caller's contract (its existing fail-safe was already correct).

### Evidence
```
$ docker compose -f docker-compose.test.yml run --rm test_studio_backend \
    python -m pytest studio/tests/app/common/routers/test_dataview_publish.py -q
............. [100%]
13 passed, 3 warnings in 0.17s
```
Backend test image built locally; the three demote/S3 tests pass alongside the existing suite.

Log emitted on a real demotion (existing logger, unchanged format):
```
Experiment {workspace_id}/{unique_id} marked synced but not present in S3
(No data found in S3 ...); demoting to error
```

### References
- Endpoint: `public_reproduce_experiment` in `studio/app/common/routers/dataview.py`
- Helper whose contract changed: `_validate_experiment_exists_in_s3`

#### Files changed
- `studio/app/common/routers/dataview.py` — add the confirm-on-S3-and-demote block to `public_reproduce_experiment`; change `_validate_experiment_exists_in_s3` to a tri-state (`True`/`False`/`None`) so transient failures are distinguishable from confirmed absence.
- `studio/tests/app/common/routers/test_dataview_publish.py` — three new tests covering demote-on-absent, no-demote-when-present, and no-demote-on-transient-error; the absent-case test also asserts the emitted `UPDATE` sets `synced → error`.

### Manual Testcases
1. **Files lost from S3** — As a viewer, request `GET /public/.../reproduce/{workspace_id}/{unique_id}` for a `synced`, published experiment whose S3 output prefix has been emptied. Expected: `503` data-error response, and the row's `local_sync_status` flips to `error` (verify in DB); the background sync job re-checks it thereafter.
2. **Transient S3 error** — Same request while S3 is briefly unreachable (`list_objects_v2` raises). Expected: `503` response, row **stays** `synced` (no DB write), retried on a later request.
3. **Data present but locally missing** — `synced` row, files present in S3 but not on local EBS. Expected: `503` response, row stays `synced` (no demotion).
4. Automated: `pytest studio/tests/app/common/routers/test_dataview_publish.py -- all 13 tests pass`

### Unit, Integration, Contract Test Coverage
- `test_reproduce_synced_but_missing_in_s3_demotes_to_error` — confirmed-absent → demote; asserts the compiled `UPDATE` sets `local_sync_status = error` under a `synced` guard.
- `test_reproduce_synced_present_in_s3_does_not_demote` — present → no `db.execute`.
- `test_reproduce_synced_s3_check_fails_does_not_demote` — transient `None` → no `db.execute`.

#### Difficulties (if any)
- None. Local conda envs lack the full backend dependency set (`stripe`, etc.), so verification was run via the Docker test image.

#### Risk Assessment
| Area | Risk | Notes |
|------|------|-------|
| `_validate_experiment_exists_in_s3` return contract | Low | Now tri-state; both call sites audited. Publish caller's `if not exists:` unchanged and still correct for the new `None`. |
| `public_reproduce_experiment` demotion write | Low | Guarded, idempotent `UPDATE`; only fires on confirmed-absent data; new branch covered by tests. |
| Background sync job interaction | Low | Demotion to `error` is the documented signal for re-check; no new coupling introduced. |

### CODE REVIEW

File: studio/tests/app/common/routers/test_dataview_publish.py [hunk 1]
```diff
@@ -473,3 +473,152 @@ async def test_reproduce_auto_updates_sync_status_when_data_available(self):
...
+                workspace_id="1", unique_id="exp123", db=mock_db
+            )
+
+        assert response.status_code == 503
+        mock_db.execute.assert_called_once()
+        mock_db.commit.assert_called_once()
+        # Pin that the statement demotes synced -> error (SET target and guard)
+        params = mock_db.execute.call_args[0][0].compile().params
+        assert params["local_sync_status"] == LocalSyncStatus.error.value
+        assert params["local_sync_status_1"] == LocalSyncStatus.synced.value
```
[the remark this answers is not in the document]
Comment (milesAraya) [reply studio/tests/app/common/routers/test_dataview_publish.py]: Addressed in [38f527e](https://github.com/arayabrain/araya-optinist/pull/748/commits/38f527ed55c6a67087bfe3e50c8c8c1249491839)

File: studio/app/common/routers/dataview.py [hunk 2]
```diff
@@ -264,6 +264,35 @@ async def public_reproduce_experiment(
...
+        if (
+            RemoteStorageController.is_available()
+            and hasattr(record, "local_sync_status")
+            and record.local_sync_status == LocalSyncStatus.synced.value
+        ):
+            bucket = _resolve_workspace_remote_bucket_name(db, workspace_id)
+            exists, s3_error = await _validate_experiment_exists_in_s3(
+                workspace_id, unique_id, bucket
+            )
+            if exists is False:
```
[the remark this answers is not in the document]
Comment (milesAraya) [reply studio/app/common/routers/dataview.py:281]: [38f527e](https://github.com/arayabrain/araya-optinist/pull/748/commits/38f527ed55c6a67087bfe3e50c8c8c1249491839)

**Review by itutu-tienday — APPROVED** [review]
I am approving this, but the AI ​​has raised a few points (regarding strengthening tests) and offered some suggestions. Please take a look at them.

---

- **Verdict:** **Approve with follow-ups.** The immediate bug fix is correct, low-risk, and ships with a genuinely valuable tri-state correctness improvement. However, it is a *targeted patch, not a generalization* — the recurrence-prevention the reviewer explicitly asked about is not achieved (see §3).

---

## 1. What the PR does

A published experiment can be `local_sync_status = synced` yet undisplayable because its S3 output files were deleted after publishing. Previously the public reproduce endpoint returned a data error and left the row stuck at `synced` forever, so the background sync job (which only scans `pending`/`error` rows) never re-checked it.

The PR:
1. In `public_reproduce_experiment`, when a `synced` row fails display validation, confirms against S3 and — **only when data is confirmed absent** — demotes the row to `error` so the background sync job re-enrolls it.
2. Changes `_validate_experiment_exists_in_s3` from a boolean to a **tri-state** return: `True` (present) / `False` (confirmed absent, `KeyCount == 0`) / `None` (couldn't check — exception).

---

## 2. Design & spec validity — sound

The core premise was verified against the codebase, not just taken from the PR description:

- **The demotion signal actually works.** `sync_job.py::_get_pending_experiments` queries published rows with `local_sync_status IN (pending, error)`. Demoting a stuck `synced` row to `error` genuinely re-enrolls it in the next validation pass. `_validate_experiment` re-checks S3 and, on persistent failure, escalates via `_check_persistent_failure`. The fix integrates correctly with the existing mechanism.
- **The tri-state refactor is a real correctness fix, not cosmetic.** Previously a transient S3 exception returned `False`. Under the new demotion logic that would have *falsely demoted a healthy row* on any network/throttle/creds blip. Splitting "confirmed absent" (`False`) from "couldn't check" (`None`) is the right abstraction, and the two fail-safe directions are correct:
  - demotion fires only on `if exists is False` — never on uncertainty (won't corrupt a healthy row);
  - publish caller keeps `if not exists:` — blocks on both `False` and `None` (won't publish unverifiable data).
- **Both call sites audited.** The publish caller (`publish_dataview_records`, ~line 560 of head) still reads `if not exists:` — behavior is unchanged for the new `None` because `not None` is truthy. Confirmed by reading the head revision.
- **Idempotent demotion.** The `UPDATE ... WHERE publish_status = on AND local_sync_status = synced` guard makes concurrent/repeat demotions converge to a no-op. Correct.

Conclusion: the spec is coherent and the fix does what it claims.

---

## 3. The reviewer's key question — *is this ad-hoc? is status management generalized so this can't recur?*

**Honest answer: it is a localized, reactive patch. Status management is NOT generalized, and this PR mildly increases the divergence.** Three concrete points:

### 3a. Detection is bolted onto a read path, not fixed at the source
The demotion only happens when a user hits `GET /public/.../reproduce/{ws}/{uid}` for that specific row. A `synced` row whose S3 data was deleted but which **nobody views** stays stuck forever. The root cause is *"`synced` rows are never re-verified against S3 data loss"* — and that lives in the background sync job, which today only validates `pending`/`error`. The general fix would be to have the sync job periodically re-verify a sample of `synced` rows; instead this PR patches one of possibly several read paths. It fixes the symptom where it hurts most, but the class of bug ("synced can silently diverge from S3") is not closed.

### 3b. Status-transition logic is scattered — there is no shared state machine
`local_sync_status` is now written in **four** places, each hand-rolling its own statement:

| Location | Mechanism | Version bump? |
|---|---|---|
| `publish_dataview_records` | optimistic-locked `UPDATE` (`WHERE version=…`) | yes |
| promote-to-synced in `public_reproduce_experiment` | optimistic-locked `UPDATE` | yes |
| **NEW** demote block | plain guarded `UPDATE` | **no** |
| `sync_job._mark_sync_complete` / `_mark_sync_error` | ORM attribute mutation | no |

There is no `SyncStatusService.demote()/promote()` (or reuse of sync_job's `_mark_*`) that all callers funnel through. This PR adds a *third* variant in the same file rather than consolidating. So to the reviewer's question directly: **no, the logic is not common/generalized, and this change adds to the sprawl.** Recommended follow-up: introduce a single transition helper (validated transition + consistent locking + logging) and route all four writers through it.

### 3c. The S3-existence check itself is duplicated
`dataview.py::_validate_experiment_exists_in_s3` and `sync_job`'s `s3_controller.validate_experiment_in_s3` are two parallel implementations of "does this experiment exist in S3." They can drift (e.g. one checks `KeyCount`, the other checks required-file lists). Consolidating onto one S3-presence primitive would reduce future divergence.

---

## 4. Code quality (lower severity)

1. **Inconsistent concurrency control.** The demote `UPDATE` omits the `version` bump that every other status write in this file performs. It relies on the `publish_status=on AND local_sync_status=synced` guard for idempotency — fine for the demotion's own correctness — but the write is invisible to optimistic-lock readers (a concurrent promote reading the same `version` would not see `rowcount == 0`). In practice the promote and demote branches are mutually exclusive within a single request, so risk is low; still, bumping `version` here would keep the file internally consistent.
2. **`s3_error` bound only inside a branch.** `s3_error` is assigned only within `if bucket:`, then read in the `if exists is False:` log. Safe *today* because `exists` can only become `False` when that branch ran (it defaults to `None`), but the invariant is implicit and fragile. Initialize `s3_error = None` alongside `exists = None` to make it robust.
3. **Redundant work.** `RemoteStorageController.is_available()` is checked in the new block and again inside `_validate_experiment_exists_in_s3`; the bucket is re-resolved via `_resolve_workspace_remote_bucket_name`. Minor.
4. **Stale in-memory record (correct, worth a note).** After the DB demotion, the in-memory `record.local_sync_status` is still `synced`; the function falls through and returns `503 data_error`. Outcome is correct, but it depends on the reader knowing the object isn't refreshed.
5. Comment and log message are clear and follow the existing format. Good.

---

## 5. Test sufficiency — adequate for the branch logic, with real gaps on the contract

**Present (good):**
- `test_reproduce_synced_but_missing_in_s3_demotes_to_error` — confirmed-absent → demote; **pins the compiled `UPDATE`** SET target (`error`) and the `synced` guard. Strong assertion.
- `test_reproduce_synced_present_in_s3_does_not_demote` — present → no `db.execute`.
- `test_reproduce_synced_s3_check_fails_does_not_demote` — transient `None` → no `db.execute`.

**Gaps:**
1. **The tri-state helper itself is never exercised.** All three tests mock `_validate_experiment_exists_in_s3`. The actual new behavior — exception → `None` (the whole point of the PR, changed from `False`), `KeyCount == 0` → `False`, else `True` — has **no direct test**. Add a unit test that drives the real helper against a mocked S3 client for each of the three outcomes. This is the single most important missing test, because the tri-state contract is the crux of the change.
2. **No regression test for the publish caller with `None`.** The PR explicitly claims `publish_dataview_records`' `if not exists:` still blocks on the new `None`. That audited-but-untested behavior is exactly what a future refactor could silently break. Add a test asserting publish returns `400` when the helper returns `(None, ...)`.
3. **No test for the `bucket` falsy branch** (bucket unresolved → `exists` stays `None` → no demote). Minor.
4. Tests use `MagicMock()` for `record`, so `hasattr(record, "local_sync_status")` is always `True`; the "record without the attribute" path is not exercised. Acceptable.

---

## 6. Recommendations

**Blocking:** none. Safe to merge as-is for the immediate incident.

**Strongly recommended (this PR or a fast follow-up):**
- Add the helper-level tri-state unit test and the publish-caller `None` regression test (§5.1, §5.2).
- Initialize `s3_error = None` for robustness (§4.2).

**Follow-up (addresses the reviewer's generalization concern — file as a tracked issue):**
- Centralize `local_sync_status` transitions behind one helper/service and route all four writers through it, with consistent locking (§3b).
- Have the background sync job periodically re-verify `synced` rows against S3, so detection no longer depends on a user hitting the read path (§3a).
- Consolidate the two S3-existence checks into one primitive (§3c).

---

*Reviewed against the PR head revision of `fix/demote-synced-experiment`; sync-job integration and the publish call site were read directly to confirm the fix's premises.*
